File: _Queue/239. Sliding Window Maximum.py
import collections

def find_max(num_list,begin=0):
    max = num_list[0]
    key = 0
    for i in range(len(num_list)):
        if num_list[i]>max:
            max = num_list[i]
            key = i+begin
    return max,key

class Solution:
    def maxSlidingWindow(self, nums, k):
        d = collections.deque()
        out = []
        for i, n in enumerate(nums):
            while d and nums[d[-1]] < n:
                d.pop()
            d += i,
            if d[0] == i - k:
                d.popleft()
            if i >= k - 1:
                out += nums[d[0]],
        return out

    def maxSlidingWindow1(self, nums, k):
        result = []
        queue = {}

        for i in range(len(nums)-k+1):

            if len(queue)==0:
                max, key = find_max(nums[i:i+k],begin=i)
                queue[key] = max
            else:
                if queue[key]>nums[i+k-1]:
                    max = queue[key]
                else:
                    queue = {}
                    max = nums[i+k-1]
                    key = i+k-1
                    queue[key] = max

            if i+1 > key:
                queue = {}

            result.append(max)

        return result

if __name__ == '__main__':
    nums,k = [1, 3, -1, -3, 5, 3, 6, 7], 3

    obj = Solution()
    print(obj.maxSlidingWindow(nums,k))

File: _Queue/cpp/239_Sliding_Window_Maximum.cpp
#include <iostream>
#include <vector>
#include <deque>
using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        deque<int> d_queue;
        for(int i = 0; i!=nums.size();i++){
            while(!(d_queue.empty()) && nums[d_queue.back()]<nums[i]){
                d_queue.pop_back();
            }
            d_queue.push_back(i);
            
            if (d_queue.front() == i - k){
                d_queue.pop_front();
            }
            
            if(i >= (k - 1)){
                result.push_back(nums[d_queue.front()]);
            }
        }
        return result;
    }
};

int main(){
    vector<int> nums = {1,-1};
    int k = 1;
    Solution obj;
    vector<int> result = obj.maxSlidingWindow(nums,k);

    for(int i = 0; i!=result.size();i++){
        cout<< result[i] << endl;
    }

    return 0;
}